### PR TITLE
channel.c: defer exit_event

### LINKED ISSUE #8813: teardown: race between init and exit_event
Sometimes the asan CI job fails like this (e.g. [here](https://github.com/neovim/neovim/pull/7813#issuecomment-380949586) and [here](https://github.com/neovim/neovim/pull/8088#issuecomment-376312511)):

```
==================== File /home/travis/build/neovim/neovim/build/log/ubsan.16820 ====================
= =================================================================
= ==16820==ERROR: AddressSanitizer: SEGV on unknown address 0x000001baab50 (pc 0x00000044f3a1 bp 0x000000000001 sp 0x7fff7ddb6a60 T0)
= ==16820==The signal is caused by a WRITE memory access.
=     #0 0x44f3a0 in atomic_compare_exchange_strong<__sanitizer::atomic_uint8_t> /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_atomic_clang.h:90:10
=     #1 0x44f3a0 in AtomicallySetQuarantineFlagIfAllocated /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_allocator.cc:520
=     #2 0x44f3a0 in __asan::Allocator::Deallocate(void*, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_allocator.cc:597
=     #3 0x4ef316 in __interceptor_free /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:50:3
=     #4 0xfbfae4 in xfree /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:134:3
=     #5 0x120e948 in free_string_option /home/travis/build/neovim/neovim/build/../src/nvim/option.c:2207:5
=     #6 0x120de3d in free_all_options /home/travis/build/neovim/neovim/build/../src/nvim/option.c:923:9
=     #7 0xfc2441 in free_all_mem /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:622:3
=     #8 0x12bbe6c in mch_exit /home/travis/build/neovim/neovim/build/../src/nvim/os_unix.c:153:3
=     #9 0x1092cc3 in exit_event /home/travis/build/neovim/neovim/build/../src/nvim/msgpack_rpc/channel.c:542:5
=     #10 0xa6b2d3 in multiqueue_process_events /home/travis/build/neovim/neovim/build/../src/nvim/event/multiqueue.c:150:7
=     #11 0xa64e2f in loop_poll_events /home/travis/build/neovim/neovim/build/../src/nvim/event/loop.c:65:3
=     #12 0x12a4ec0 in os_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/os/input.c:162:5
=     #13 0x1035f07 in line_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/misc1.c:2692:5
=     #14 0x143ac50 in nfa_regmatch /home/travis/build/neovim/neovim/build/../src/nvim/regexp_nfa.c:6203:5
=     #15 0x141b4d9 in nfa_regtry /home/travis/build/neovim/neovim/build/../src/nvim/regexp_nfa.c:6272:16
=     #16 0x1419203 in nfa_regexec_both /home/travis/build/neovim/neovim/build/../src/nvim/regexp_nfa.c:6440:12
=     #17 0x13c734d in nfa_regexec_nl /home/travis/build/neovim/neovim/build/../src/nvim/regexp_nfa.c:6593:10
=     #18 0x1379f20 in vim_regexec_both /home/travis/build/neovim/neovim/build/../src/nvim/regexp.c:7250:16
=     #19 0x137ac07 in vim_regexec_nl /home/travis/build/neovim/neovim/build/../src/nvim/regexp.c:7302:10
=     #20 0x910f2b in f_split /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15673:17
=     #21 0x829124 in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6368:11
=     #22 0x83d018 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6114:11
=     #23 0x8c57a1 in eval7 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:4218:15
=     #24 0x8c1d6b in eval6 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3915:7
=     #25 0x8bfa7f in eval5 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3766:7
=     #26 0x8bb29f in eval4 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3503:7
=     #27 0x8ba73c in eval3 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3421:7
=     #28 0x8b9bdc in eval2 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3352:7
=     #29 0x8212f1 in eval1 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3279:7
=     #30 0x82071b in eval0 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3239:9
=     #31 0x82d088 in ex_let /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:1485:9
=     #32 0xb69fc3 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2228:5
=     #33 0xb4c569 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:592:20
=     #34 0x85cbc5 in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21442:3
=     #35 0x828a4b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6353:11
=     #36 0x83d018 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6114:11
=     #37 0x835fd4 in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2727:9
=     #38 0xb69fc3 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2228:5
=     #39 0xb4c569 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:592:20
=     #40 0xd2d506 in apply_autocmds_group /home/travis/build/neovim/neovim/build/../src/nvim/fileio.c:6933:5
=     #41 0xcf8ff6 in apply_autocmds /home/travis/build/neovim/neovim/build/../src/nvim/fileio.c:6576:10
=     #42 0x1168c78 in normal_check_text_changed /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:1219:5
=     #43 0x11667d5 in normal_check /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:1325:5
=     #44 0x17774c1 in state_enter /home/travis/build/neovim/neovim/build/../src/nvim/state.c:28:35
=     #45 0x10c0bf4 in normal_enter /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:467:3
=     #46 0xea98a2 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:566:3
=     #47 0x7fbb0a667f44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
=     #48 0x44e39b in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x44e39b)
= 
= AddressSanitizer can not provide additional info.
= SUMMARY: AddressSanitizer: SEGV /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_atomic_clang.h:90:10 in atomic_compare_exchange_strong<__sanitizer::atomic_uint8_t>
= ==16820==ABORTING
=====================================================================================================
test/helpers.lua:149: assertion failed!
```

Possible cause:

-  Options init isn't complete, so `free_all_options` is free'ing something before it was alloc'd.
- stdio RPC channel was closed
    - was this because the client died from a invalid message?
-  `exit_event` is scheduled on `main_loop.fast_events`. 
    - is this necessary, or could we put it on `main_loop.events` instead? Then `os_breakcheck .. loop_poll_events`

### PR DESCRIPTION
**Pull request by justinmk**
closes #8813
ref https://github.com/neovim/neovim/pull/7813#issuecomment-380949586
ref https://github.com/neovim/neovim/pull/8088#issuecomment-376312511

Deferring the event:

```diff
diff --git a/src/nvim/msgpack_rpc/channel.c b/src/nvim/msgpack_rpc/channel.c
index 3244d83a93c..62c7717c2e9 100644
--- a/src/nvim/msgpack_rpc/channel.c
+++ b/src/nvim/msgpack_rpc/channel.c
@@ -532,7 +532,7 @@ void rpc_close(Channel *channel)
   channel_decref(channel);
 
   if (channel->streamtype == kChannelStreamStdio) {
-    multiqueue_put(main_loop.fast_events, exit_event, 0);
+    multiqueue_put(main_loop.events, exit_event, 0);
   }
 }
```

...leads to hangs (waiting for input):
 
```
[5/7] Building C object test/functiona...s/CMakeFiles/tty-test.dir/tty-test.c.o[K
[6/7] Linking C executable bin/tty-test[K
[6/7] Linking C executable bin/tty-test[K
[7/7] cd /home/travis/build/neovim/neo...ild/neovim/neovim/cmake/RunTests.cmake[K

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

---

Notes from chat with @bfredl :

> The problem is that scheduling an event doesn't guarantee any progress back to the main normal mode.

* Screen tests may hang (regardless of success or failure), after the test harness (busted) closes the channel, but nvim is stuck in press-enter
* Root cause is that Nvim doesn't check for events during press-enter (`events_enabled=0` in `os/input.c`)
    * `normal_execute..wait_return..os_inchar` waits for input forever, because it doesn't process `main_loop.events`, because `events_enabled=0`, because our current policy is that it's unsafe to process `main_loop.events` until we get back to the top of the `state_enter` loop. See also https://github.com/neovim/neovim/pull/8814#issuecomment-422274556

### CODE REVIEW

File: src/nvim/msgpack_rpc/channel.c [hunk 1]
```diff
@@ -539,7 +540,9 @@ void rpc_close(Channel *channel)
 static void exit_event(void **argv)
 {
   if (!exiting) {
-    mch_exit(0);
+    // <C-\><C-N>:qall!<CR>
+    ins_typebuf((char_u *)"\034\016:qall!\n", REMAP_NONE, typebuf.tb_len, 1,
```
[the remark this answers is not in the document]
Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c]: @bfredl Thanks for taking a look. Do you mean:

    multiqueue_put(main_loop.events, exit_event, 0);

that hangs (see PR description). But this `ins_typebuf()` also hangs, probably for the same reason.

[the remark this answers is not in the document]
Comment (bfredl) [reply src/nvim/msgpack_rpc/channel.c]: Where does it hang? Possible to get a traceback?

**Comment by bfredl** [comment]
Or should we just skip `free_all_mem` when nvim exits in a breakcheck? This is an artificial condition anyway not related to any real-world use.

**Comment by bfredl** [comment]
Remember, vim is _by design_ not gc-safe in any place except for top-level, so I don't see how free_all_mem could be.

**Comment by justinmk** [comment]
> Or should we just skip free_all_mem when nvim exits in a breakcheck?

But then we'd just get a different kind of false positive in the asan job, no?

File: src/nvim/msgpack_rpc/channel.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c]: workin' on it

**Comment by bfredl** [comment]
I meant with some kind of message that a leak will be indeterministically (rarely) expected. A leak is a more reasonable failure than a crash that won't ever happen in real usage.

**Comment by justinmk** [comment]
> Where does it hang? Possible to get a traceback?

It's stuck in `loop_poll_events`. And that makes sense, because `loop_poll_events` doesn't check `Loop.events`. But this also means that if we schedule something on `main_loop.events` it's not guaranteed to be processed until input (`Loop.uv`) is received.

Normally we don't want `loop_poll_events` to check `Loop.events` because that's application-specific. But I wonder if we should timeout `loop_poll_events` so that `normal_enter` occasionally checks `main_loop.events`.

```
(gdb) bt
+bt
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000000000074fe52 in uv__io_poll (loop=<optimized out>, timeout=-1) at /home/vagrant/neovim/.deps/build/src/libuv/src/unix/linux-core.c:289
#2  0x0000000000742c64 in uv_run (loop=0xac5bd0 <main_loop>, mode=UV_RUN_ONCE) at /home/vagrant/neovim/.deps/build/src/libuv/src/unix/core.c:359
#3  0x00000000004ffbf3 in loop_poll_events (loop=0xac5bd0 <main_loop>, ms=-1) at ../src/nvim/event/loop.c:58
#4  0x0000000000627d7a in input_poll (ms=-1) at ../src/nvim/os/input.c:354
#5  0x0000000000626d90 in inbuf_poll (ms=-1) at ../src/nvim/os/input.c:377
#6  0x0000000000626cae in os_inchar (buf=0x0, maxlen=0, ms=-1, tb_change_cnt=0) at ../src/nvim/os/input.c:120
#7  0x00000000006cef1f in state_enter (s=0x7ffe4a3fe838) at ../src/nvim/state.c:55
#8  0x00000000005e4dc4 in normal_enter (cmdwin=false, noexmode=false) at ../src/nvim/normal.c:467
#9  0x00000000005a120f in main (argc=9, argv=0x7ffe4a3febd8) at ../src/nvim/main.c:566
```

**Comment by bfredl** [comment]
That doesn't make sense and is a bug, ordinary normal mode should always process main_loop.events before waiting on input.

**Comment by justinmk** [comment]
> That doesn't make sense and is a bug, ordinary normal mode should always process main_loop.events before waiting on input.

@bfredl  It does. But the stream might be closed _during_ the hit-enter prompt, after `events_enabled=0`. That's the problem: RPC events are not processed during hit-enter.

* We can deal with this in the test harness by sending CTRL-C or something like that before closing the stream. 
* Should we try a more robust solution which makes life easier for UI clients? 
    * Special-case the `exit_event` case? 
    * Process RPC events during press-enter?

**Comment by bfredl** [comment]
> Process events during press-enter?

I would vote for this. clients shouldn't randomly do potentially dangerous stuff in an async context (i e not as part of handling an event from nvim) anyway, they can use `nvim_get_mode` and check the mode. Perhaps there should even be a mechanism for a client to ask "please send me a request as soon as nvim is back to top-level normal mode".

File: src/nvim/os/input.c [hunk 2]
```diff
@@ -88,9 +88,9 @@ static void create_cursorhold_event(void)
 {
   // If events are enabled and the queue has any items, this function should not
   // have been called(inbuf_poll would return kInputAvail)
-  // TODO(tarruda): Cursorhold should be implemented as a timer set during the
+  // TODO(justinmk): Cursorhold should be implemented as a timer set during the
   // `state_check` callback for the states where it can be triggered.
-  assert(!events_enabled || multiqueue_empty(main_loop.events));
+  assert(multiqueue_empty(main_loop.events));
```
[the remark this answers is not in the document]
Comment (bfredl) [reply src/nvim/os/input.c]: Why is that? is K_EVENT regarded as a dismissive key in `wait_return`? Maybe it needs to be added to the list at `messages.c:954-966` of "allowed" state r keys.

**Pull request closed without merging**